rts/motoko-rts/src/gc/generational/sanity_checks.rs

```diff
@@ -0,0 +1,183 @@
+        panic!("Missing write barrier at {:#x}", current_field as usize);
+    }
+}
+
+unsafe fn recorded(value: u32) -> bool {
+    match &REMEMBERED_SET {
+        None => panic!("No remembered set"),
+        Some(remembered_set) => {
+            let mut iterator = remembered_set.iterate();
+            while iterator.has_next() {
+                if iterator.current().get_raw() == value {
+                    return true;
+                }
+                iterator.next();
+            }
+            false
+        }
+    }
+}
+
+pub struct MemoryChecker<'a> {
+    limits: &'a Limits,
+    roots: &'a Roots,
+}
+
+pub unsafe fn check_memory(limits: &Limits, roots: &Roots) {
+    let checker = MemoryChecker { limits, roots };
+    checker.check_memory();
+}
+
+impl<'a> MemoryChecker<'a> {
+    unsafe fn check_memory(&self) {
+        self.check_static_roots();
+        if (*self.roots.continuation_table_ptr_loc).is_ptr() {
+            self.check_object(*self.roots.continuation_table_ptr_loc);
+        }
+        self.check_heap();
+    }
+
+    unsafe fn check_static_roots(&self) {
+        let root_array = self.roots.static_roots.as_array();
+        for i in 0..root_array.len() {
+            let obj = root_array.get(i).as_obj();
+            assert_eq!(obj.tag(), TAG_MUTBOX);
+            assert!((obj as usize) < self.limits.base);
+            let mutbox = obj as *mut MutBox;
+            let field_addr = &mut (*mutbox).field;
+            if pointer_to_dynamic_heap(field_addr, self.limits.base as usize) {
+                let object = *field_addr;
+                self.check_object(object);
+            }
+        }
+    }
+
+    unsafe fn check_object(&self, object: Value) {
+        self.check_object_header(object);
+        visit_pointer_fields(
+            &mut (),
+            object.as_obj(),
+            object.tag(),
+            0,
+            |_, field_address| {
+                if Self::is_ptr(*field_address) {
+                    (&self).check_object_header(*field_address);
+                }
+            },
+            |_, _, arr| arr.len(),
+        );
+    }
+
+    unsafe fn check_object_header(&self, object: Value) {
+        assert!(Self::is_ptr(object));
+        let pointer = object.get_ptr();
+        assert!(pointer < self.limits.free);
+        let tag = object.tag();
+        assert!(tag >= TAG_OBJECT && tag <= TAG_NULL);
+    }
+
+    unsafe fn check_heap(&self) {
+        let mut pointer = self.limits.base;
+        while pointer < self.limits.free {
+            let object = Value::from_ptr(pointer as usize);
+            if object.tag() != TAG_ONE_WORD_FILLER {
+                self.check_object(object);
+            }
+            pointer += object_size(pointer as usize).to_bytes().as_usize();
+        }
+    }
+
+    unsafe fn is_ptr(value: Value) -> bool {
+        const TRUE_VALUE: u32 = 1;
+        value.is_ptr() && value.get_raw() != TRUE_VALUE
```

Comment:
Should value.is_ptr() not filter out TRUE_VALUE itself?

Reply:
I moved this logic. First, I was unsure what impact it has on the remaining code (also performance-wise). I checked this and it seems to be good.

Reply:
I optimized this in [6fc0079](https://github.com/dfinity/motoko/pull/3495/commits/6fc0079c5696b287e538026cbd2f399c4d379ce9). A similar optimization was done in compacting GC for comparing the field addresses. So, performance is as good as it was before this change.